src/audio/kpb.c

```diff
@@ -157,8 +161,178 @@ static void kpb_lock_init(struct comp_data *kpb)
+ * \param[in] dev - component device pointer.
+ * \param[in] ipc_config - ipc config pointer.
+ * \return: none.
+ */
+static int kpb_set_verify_ipc_params(struct comp_dev *dev,
+				     const struct ipc4_kpb_module_cfg *ipc_config)
+{
+	struct comp_data *kpb;
+
+	kpb->config.channels = ipc_config->base_cfg.audio_fmt.channels_count;
+	kpb->config.sampling_freq =
+		ipc_config->base_cfg.audio_fmt.sampling_frequency;
+	kpb->config.sampling_width =
+		ipc_config->base_cfg.audio_fmt.valid_bit_depth;
+	kpb->ipc4_cfg.base_cfg = ipc_config->base_cfg;
+
+	/* Initialize sinks */
+	kpb->sel_sink = NULL;
+	kpb->host_sink = NULL;
+
+	if (!kpb_is_sample_width_supported(kpb->config.sampling_width)) {
+		comp_err(dev, "kpb_set_verify_ipc_params(): requested sampling width not supported");
+		return -EINVAL;
+	}
+
+	if (kpb->config.channels > KPB_MAX_SUPPORTED_CHANNELS) {
+		comp_err(dev, "kpb_set_verify_ipc_params(): no of channels exceeded the limit");
+		return -EINVAL;
+	}
+
+	if (kpb->config.sampling_freq != KPB_SAMPLNG_FREQUENCY) {
+		comp_err(dev, "kpb_set_verify_ipc_params(): requested sampling frequency not supported");
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+/**
+ * \brief Set KPB component stream params.
+ * \param[in] dev - component device pointer.
+ * \param[in] params - sof ipc strem params pointer.
+ * \return: none.
+ */
+static void kpb_set_params(struct comp_dev *dev,
+			   struct sof_ipc_stream_params *params)
+{
+	struct comp_data *kpb = comp_get_drvdata(dev);
+	uint32_t __sparse_cache valid_fmt, frame_fmt;
+
+	comp_dbg(dev, "kpb_set_params()");
+
+	memset_s(params, sizeof(*params), 0, sizeof(*params));
+	params->channels = kpb->ipc4_cfg.base_cfg.audio_fmt.channels_count;
+	params->rate = kpb->ipc4_cfg.base_cfg.audio_fmt.sampling_frequency;
+	params->sample_container_bytes = kpb->ipc4_cfg.base_cfg.audio_fmt.depth / 8;
+	params->sample_valid_bytes =
+		kpb->ipc4_cfg.base_cfg.audio_fmt.valid_bit_depth / 8;
+	params->buffer_fmt = kpb->ipc4_cfg.base_cfg.audio_fmt.interleaving_style;
+	params->buffer.size = kpb->ipc4_cfg.base_cfg.obs * KPB_MAX_BUFF_TIME * 2;
+
+	params->host_period_bytes = params->channels *
+				    params->sample_container_bytes *
+				    (params->rate / 1000);
+
+	audio_stream_fmt_conversion(kpb->ipc4_cfg.base_cfg.audio_fmt.depth,
+				    kpb->ipc4_cfg.base_cfg.audio_fmt.valid_bit_depth,
+				    &frame_fmt, &valid_fmt,
+				    kpb->ipc4_cfg.base_cfg.audio_fmt.s_type);
+
+	params->frame_fmt = frame_fmt;
+}
+
+/**
+ * \brief Set KPB component stream params.
+ * \param[in] dev - component device pointer.
+ * \param[in] type - sof ipc stream params pointer.
+ * \param[in] value - ipc4 base module config pointer.
+ * \return: none.
+ */
+static int kpb_get_attribute(struct comp_dev *dev,
+			     uint32_t type,
+			     void *value)
+{
+	struct comp_data *kpb = comp_get_drvdata(dev);
+
+	switch (type) {
+	case COMP_ATTR_BASE_CONFIG:
+		*(struct ipc4_base_module_cfg *)value = kpb->ipc4_cfg.base_cfg;
```

Comment:
this wont align with the module adapter's get_atrribute() callback and will make it difficult to convert

Reply:
@ranj063 what change is needed here ?